MetaCell/salk-interactive-atlas
# Feature/connect UI

Closes #32 #47 #49 #51

- Splits salk GM and WM meshes per segment. Script used to do so added in scripts folder
- Mocks get experiment and misc (all mocked methods should start with MOCKED).
- Creates SimpleInstances for all the atlas meshes.
- Adds `Atlas`, `AtlasInstance`, `AtlasSegment`, and `Range` as frontend models
- Implements subdivision and population sliders logic.

Implemented solution: 

**- Handle of Atlas Static Data:**

The Atlas are generated at runtime by dynamically iterating over the `assets/atlas` folder (see constants.ts and Atlas.ts)
It is expected that atlas static data to be stored in the aforementioned folder.
This data should be wrapped in a folder that need to match with the ids in the `AtlasChoice` enum (see constants.ts)
The meshes inside the folder should have as filename `{Depth}_{Segment}_{StructureName}.obj`
Each obj file results in a SimpleInstance object (see AtlasInstances.ts)
The segments information should be stored aside the meshes with the name `atlas_segments.json`

**- General Components Flow:**

`ExperimentPage` connects both the `ExperimentViewer` and the `ExperimentSidebar`.

`ExperimentSidebar` receives `selectedAtlas`, `subdivisions` and `populations` as props as well as all the handlers to manage the sliders
          `SelectedAtlas` starts with  `AtlasChoice.slk10` as default
          `Subdivisions` come from the static atlas
          `Populations` come from the mocked experiment.
          Handlers manage the `selected` attribute of the aforementioned entities.
          
`ExperimentViewer` receives `selectedAtlas`, `activeSubdivisions` and `activePopulations` as props. 
          `activeSubdivisions` is a set of ids of the active subdivisions
          `activePopulations` is a map<PopulationId, PopulationEntity>


**- `ExperimentViewer` Detailed implementation:**

We iterate over the (selected) Atlas instances and check if they are from any of the `activeSubdivisions`. If they are, we consider them for the Canvas proxy instances. Color information is still MOCKED.

The populations are handled outside the Canvas component. We get access to the scene created by the canvas and interact with it to create multiple `SphereGeometry` using `InstancedMesh`.
We check if the X component of the `Cell` is inside any range of the `activeSubdivisions`.

Updates on `activeSubdivisions` do trigger an update on the Canvas because the `proxyInstances` change.
It will also result in a clean (remove all cells and add again) update in the populations.

Updates on the `activePopulations` will only remove or add populations if needed. We check the different between currentState and desiredState.
          
How to test this PR: 

See video attached for how to trigger the features added

Known Issues to be address with Geppetto changes:
- Flip camera https://github.com/MetaCell/geppetto-meta/issues/321
- Reset Logic (In the video I click on pan home to bring the camera to a nice position after adding the meshes, that shouldn't be necessary) https://github.com/MetaCell/geppetto-meta/issues/338
- Material Opacity and render order https://github.com/MetaCell/geppetto-meta/issues/297


### Sanity checks:

### Breaking changes (select one):

### Possible deployment updates issues (select one):

### Test coverage (select one):

### Documentation (select one):

### Nice to have (if relevant):


https://user-images.githubusercontent.com/19196034/158714788-9c648efb-4a68-4f65-ba81-d5727bf1e4b4.mp4
